[OWL-456] Implement nqmRpc mechanism, which support Cassandra

humorless opened the pull request:
[OWL-456] Implement nqmRpc mechanism, which support communication with Cassandra.
## How to test:

1 Run the transfer
2 Simulate pkt from NQM agent by 
   cd test
   debug nqm
3 Check the jsonrpc2 pkt sent out from transfer by
   netcat -l  [port]
   Or  using the jsonRpc2 server  deliberately written for this. See the attachment. 
[main.go.txt](https://github.com/Cepave/transfer/files/223416/main.go.txt)

humorless committed:
[OWL-456] Implement nqmRpc mechanism, which support communication with Cassandra.

humorless committed:
[OWL-456] makes nqmRpc's URL can accept path string

Before this modification:
    "nqmRpc": {
        "enabled":true,
        "batch": 100,
        "connTimeout": 1000,
        "callTimeout": 5000,
        "maxConns": 32,
        "maxIdle": 32,
        "addr": "10.0.0.135:8080"
    }

After this small modification, URL can accept path string
    "nqmRpc": {
        "enabled":true,
        "batch": 100,
        "connTimeout": 1000,
        "callTimeout": 5000,
        "maxConns": 32,
        "maxIdle": 32,
        "addr": "10.0.0.135:8080/jsonrpc"
    }

humorless committed:
[OWL-456] Add the pkt demultiplexing mechanism

Nqm pkt, which metric as the value "nqm-metrics" will be sent to
Cassandra.
Generic pkt, which metric has the value not equal to "nqm-metrics"
will be sent to other databases.

mikelue commented on sender/sender.go:
```diff
@@ -227,3 +230,166 @@ func Push2InfluxdbSendQueue(items []*cmodel.MetaData) {
+		if err != nil {
+			return &t, err
+		}
+	}
```
1. The `if err != nil` statement should be put after `ii, err = strconv.ParseInt(v, 10, 32)` immediately.
2. While **error is not nil**, use `nil` as return value would be better
3. If you like more information for error context, try using `fmt.Errorf()` to generate detail information.

For example:

```
v, err = someFunc()
if err != nil {
    return nil, fmt.Errorf("Your error context: %v", err)
}
```

humorless committed:
[OWL-456] refactor convert2NqmRpcItem under unit test.

convert2NqmRpcItem calls
(1) convert2NqmEndpoint
(2) convert2NqmMetrics

In this commit, refactor the two functions called by
convert2NqmRpcItem. Enhancing the exception handling part
to let error be easily discovered.

chyeh commented on g/git.go:
```diff
@@ -1,4 +1,4 @@
 package g
 const (
-    COMMIT = "1ee55c3"
+    COMMIT = "bc022cd"
```
Is this necessary?

chyeh commented on sender/conn_pool/nqmrpc_conn_pool_manager.go:
```diff
@@ -0,0 +1,100 @@
+	return this.cli == nil
+}
+
+func (this RpcClient2) Close() error {
```
It would be better to use a more specific name rather than `RpcClient2`.

humorless committed:
[OWL-456] small modification by suggestions from comments

1 slightly modify the strToInt32, strToInt16, strToFloat32.
  map is reference type and it will be treated as call by
  reference.

2 fix the commit value in git.go
3 fix the type name from RpcClient2 to JsonRpcV2Client

humorless committed:
[OWL-456] Modified because expected Rttmin and Rttmax will be float.

humorless committed:
[OWL-456] Modified the JSONRPC2 library from powerman

In powerman implementation, it has very strict http content type
checking, which makes it difficult to communicate with other
JSONRPC2 implementation. Therefore, I just remove the http
content type checking.

humorless committed:
[OWL-456] Change the jsonrpc2 library import path

    Before, transfer module modified jsonrpc2 implementation.

    In this commit, move the modified jsonrpc2 implementation to
    http://github.com/Cepave/rpc-codec
